src/cpp/common/alarm.cc

```diff
@@ -52,78 +62,87 @@ class AlarmImpl : public grpc::internal::CompletionQueueTag {
     cq_ = cq->cq();
     tag_ = tag;
     GPR_ASSERT(grpc_cq_begin_op(cq_, this));
-    GRPC_CLOSURE_INIT(
-        &on_alarm_,
-        [](void* arg, grpc_error_handle error) {
-          // queue the op on the completion queue
-          AlarmImpl* alarm = static_cast<AlarmImpl*>(arg);
-          alarm->Ref();
-          // Preserve the cq and reset the cq_ so that the alarm
-          // can be reset when the alarm tag is delivered.
-          grpc_completion_queue* cq = alarm->cq_;
-          alarm->cq_ = nullptr;
-          grpc_cq_end_op(
-              cq, alarm, error,
-              [](void* /*arg*/, grpc_cq_completion* /*completion*/) {}, arg,
-              &alarm->completion_);
-          GRPC_CQ_INTERNAL_UNREF(cq, "alarm");
-        },
-        this, grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&timer_,
-                    grpc_core::Timestamp::FromTimespecRoundUp(deadline),
-                    &on_alarm_);
+    Ref();
+    cq_armed_.store(true);
+    cq_timer_handle_ = event_engine_->RunAfter(
+        grpc_core::Timestamp::FromTimespecRoundUp(deadline) -
+            grpc_core::ExecCtx::Get()->Now(),
+        [this] {
+          cq_armed_.store(false);
+          OnCQAlarm(absl::OkStatus());
+        });
   }
   void Set(gpr_timespec deadline, std::function<void(bool)> f) {
-    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
     // Don't use any CQ at all. Instead just use the timer to fire the function
     callback_ = std::move(f);
     Ref();
-    GRPC_CLOSURE_INIT(
-        &on_alarm_,
-        [](void* arg, grpc_error_handle error) {
-          grpc_core::Executor::Run(GRPC_CLOSURE_CREATE(
-                                       [](void* arg, grpc_error_handle error) {
-                                         AlarmImpl* alarm =
-                                             static_cast<AlarmImpl*>(arg);
-                                         alarm->callback_(error.ok());
-                                         alarm->Unref();
-                                       },
-                                       arg, nullptr),
-                                   error);
-        },
-        this, grpc_schedule_on_exec_ctx);
-    grpc_timer_init(&timer_,
-                    grpc_core::Timestamp::FromTimespecRoundUp(deadline),
-                    &on_alarm_);
+    callback_armed_.store(true);
+    callback_timer_handle_ = event_engine_->RunAfter(
+        grpc_core::Timestamp::FromTimespecRoundUp(deadline) -
+            grpc_core::ExecCtx::Get()->Now(),
+        [this] {
+          callback_armed_.store(false);
+          OnCallbackAlarm(true);
+        });
   }
   void Cancel() {
-    grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
     grpc_core::ExecCtx exec_ctx;
-    grpc_timer_cancel(&timer_);
+    if (callback_armed_.load() &&
+        event_engine_->Cancel(callback_timer_handle_)) {
+      event_engine_->Run([this] { OnCallbackAlarm(/*is_ok=*/false); });
```

Comment:
should we set callback_armed_.store(false) inside the if statement ?

Reply:
Yes indeed, good catch. That means we don't have any tests that exercise a successful cancellation path with multi-set on the same Alarm. I'll add a test.

Reply:
Added some failing tests, then fixed the implementation.